src/components/Tooltip.js

```diff
@@ -0,0 +1,135 @@
+
+const propTypes = {
+    // The text to display in the tooltip.
+    text: PropTypes.string.isRequired,
+
+    // Child to wrap with Tooltip. Must be only child.
+    children: PropTypes.oneOfType([
+        PropTypes.func,
+        PropTypes.element,
+        PropTypes.instanceOf(Component),
+    ]).isRequired,
+};
+
+class Tooltip extends Component {
+    constructor(props) {
+        super(props);
+
+        this.animation = new Animated.Value(0);
+
+        // The wrapper view containing the wrapped content along with the Tooltip itself.
+        this.wrapperView = null;
+        this.width = 0;
+        this.height = 0;
+
+        // The tooltip (popover) itself.
+        this.tooltip = null;
+        this.tooltipWidth = 0;
+        this.tooltipHeight = 0;
+
+        // The distance between the left side of the rendered view and the left side of the window
+        this.xOffset = 0;
+
+        // The distance between the top of the rendered view and the top of the window
+        this.yOffset = 0;
+
+        this.getPosition = this.getPosition.bind(this);
+        this.showTooltip = this.showTooltip.bind(this);
+        this.hideTooltip = this.hideTooltip.bind(this);
+    }
+
+    /**
+     * Get the position and measure the size of the wrapper view and the tooltip itself.
+     */
+    getPosition() {
+        this.wrapperView.measureInWindow((x, y, width, height) => {
+            this.xOffset = x;
+            this.yOffset = y;
+            this.width = width;
+            this.height = height;
+        });
+        this.tooltip.measureInWindow((x, y, width, height) => {
+            this.tooltipWidth = width;
+            this.tooltipHeight = height;
+        });
+    }
+
+    /**
+     * Display the tooltip in an animation.
+     */
+    showTooltip() {
+        Animated.timing(this.animation, {
+            toValue: 1,
+            duration: 140,
+        }).start();
+    }
+
+    /**
+     * Hide the tooltip in an animation.
+     */
+    hideTooltip() {
+        Animated.timing(this.animation, {
+            toValue: 0,
+            duration: 140,
+        }).start();
+    }
+
+    render() {
+        const interpolatedSize = this.animation.interpolate({
+            inputRange: [0, 1],
+            outputRange: [0, 1],
+        });
+        const {
+            animationStyle,
+            tooltipWrapperStyle,
+            tooltipTextStyle,
+            pointerWrapperStyle,
+            pointerStyle,
+        } = getTooltipStyles(
+            interpolatedSize,
+            this.xOffset,
+            this.yOffset,
+            this.width,
+            this.height,
+            this.tooltipWidth,
+            this.tooltipHeight,
+        );
+
+        return (
+            <Hoverable
+                onHoverIn={this.showTooltip}
+                onHoverOut={this.hideTooltip}
+            >
+                <View
+                    ref={el => this.wrapperView = el}
+                    onLayout={this.getPosition}
```

Comment:
Does `onLayout` return layout props we can use? https://reactnative.dev/docs/view#onlayout

Reply:
Sorry, not sure I understand your question...

From what I can tell, `onLayout` provides its callback 4 arguments:

1. `x` -> the distance between the left side of the window and the left side of the view
2. `y` -> the distance between the top of the window and the top of the view
3. `width` -> the width of the view
4. `height` -> the height of the view

Reply:
Right, so we then call `getPosition()` which seems to return the same stuff when it calls `measureInWindow()` ?

https://reactnative.dev/docs/direct-manipulation#measureinwindowcallback

So another way to phrase my question would be... why do we need `measureInWindow()` and can we just use the layout props returned by `onLayout`?

Reply:
Ahh, I remember why. `onLayout` is relative to the parent element, but `measureInWindow` gives you the position relative to the window, which is what we want in this case I think.

https://github.com/facebook/react-native/issues/10556

See how they're different:

![image](https://user-images.githubusercontent.com/47436092/105921375-9f557d80-5fed-11eb-83cf-60f7418e7ada.png)

Reply:
So my above comment wasn't correct. `onLayout` gives you:

1) `x` -> the distance between the left side of the parent and the left side of the view
2) `y` -> the distance between the top of the parent and the top of the view.

Reply:
Added a comment in the code explaining why we use `measureInWindow`